Bowling.java

```diff
@@ -116,28 +120,50 @@ private Frame getCurrentFrame() {
 		return currentFrame;
 	}
 	
-	private Optional<Frame> previousFrame() {
-		if (allFrames.size() > 1) {
-			int nextToLastIndex = allFrames.size() - 2;
+	private Optional<Frame> getFrameBefore(Frame frame) {
+		int currentIndex = allFrames.indexOf(frame);
+		int lastIndex = allFrames.size() - 1;
+
+		if (allFrames.size() > 1 && currentIndex == lastIndex) {
+			int nextToLastIndex = allFrames.indexOf(frame) - 1;
 			return Optional.of(allFrames.get(nextToLastIndex));
+		} else if (allFrames.size() > 2) {
+			int thirdToLastIndex = allFrames.size() - 3;
+			return Optional.of(allFrames.get(thirdToLastIndex));
 		}
```

Comment:
Yes, this was not well organized. 

I changed it so that `nextToLastIndex` and `thirdToLastIndex` variables are defined up front and included the `currentIndex` in the second condition, which I hope makes it read better. 

I think we need both parts of the condition: the first part (checks the size of `allFrames` is a certain size) is necessary so that we don't ask `allFrames` for an element at an index that doesn't exist, and the second part (compares `currentIndex` to that of the last element or second-to-last element) is necessary in order to know which frame `getFrameBefore` should go to and then give us the frame before.

Perhaps there is a different way of structuring this entirely if this still doesn't feel right...